### DISCUSSION TITLE
Best strategy to tackle many npm dependencies and longish pipeline execution time

### DISCUSSION THREAD

**Discussion by florianklumb** [renovatebot/renovate#15376] [Request Help] [answered]
I have a `package.json` with around 120 dependencies (most of them `devDependencies` but still). The build and test pipeline takes around 20 minutes per run. If I use the default config, I soon get 10 PRs which run and ideally pass the pipeline. Let's assume RenovateBot is set to automerge. The first PR would be merged, but then 9 remaining branches would be rebased and the pipeline would be executed a second time for these 9 branches and so on. So for these 10 PRs I get 10+9+...+1 = 55 runs which means around 
1100 minutes of pipeline time where in fact only 200 minutes would have been needed. Because every time a PR is merged a new PR is opened, every branch is built and tested 10x on average - which is not good. What can I do? 

My ideas:
- Yes, I probably should do more grouping - I'll look into that next week.
- Run RenovateBot often but only allow 1-2 concurrent PRs. Disadvantage: RenovateBot also needs ressources, crazy amount of log files.
- Disable patch level updates (maybe except 0.0.x ones).

Any other ideas?

**Comment by rarkins** [comment, marked as answer]
I don't think the disadvantages of point 2 makes logical sense. Reducing concurrent PRs reduces resources substantially. Even if it's still high in your opinion, that's not a disadvantage of point 2 itself

**Comment by viceice** [comment]
if you're just starting with renovate you should reduce concurrent branches to a low value until your repo is uptodate.

you should also only rebase renovate branches when they are conflicted or for automerge.

**Comment by PhilipAbed** [comment]
It's possible to restrict concurrent PRs by config when using: [prconcurrentlimit](
https://docs.renovatebot.com/configuration-options/#prconcurrentlimit)
